### PR TITLE
async.js tests

### PR DESCRIPTION
**Pull request by marcoscaceres**

### CODE REVIEW

**Comment by marcoscaceres** [comment]
screenshot_2015-08-17_21_55_24

**Comment by oyiptong** [comment]
Can you please move async.js into /js/lib/async.js?
Also, how about we make the test directory structure mirror the code structure?
i.e. ./test/lib/async.js?

File: test/async_test.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    });
+    return test().should.eventually.be.rejectedWith(error);
+  });
+
+  it("should recover and return in a catch.", () => {
+    var error = new Error("");
+    var test = async(function* () {
+      try {
+        yield Promise.reject(error);
+      } catch (err) {
```
Comment (oyiptong) [inline test/async_test.js]: Neat new feature!

File: test/async_test.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        yield Promise.reject(new Error("pass"));
+      } catch (error) {
+        result = error.message;
+      }
+      return result;
+    });
+    return test().should.become("pass");
+  });
+
+  it("should recover from rejection.", () => {
```
Comment (oyiptong) [inline test/async_test.js]: want to use another description than from line 44?

File: test/async_test.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        err = ex.message;
+      }
+      var w = yield (err + "recovered_");
+      var z = yield (w + "123");
+      return z; //exception_recovered_123
+    });
+    return test().should.become("exception_recovered_123");
+  });
+
+  it("should recover from rejection, then resolve with a string", () => {
```
Comment (oyiptong) [inline test/async_test.js]: What is the difference from the test in line 57? the fact that it's calling Promise.resolve?

**Comment by oyiptong** [comment]
needs a test that tests line 40, when a function is not passed to async

File: test/async_test.js (hunk 3 above)
Comment (marcoscaceres) [inline test/async_test.js]: Yep. Not a big difference. But it passing a promise back uses a different part of the underlying machinery.

File: test/async_test.js (hunk 1 above)
Comment (marcoscaceres) [inline test/async_test.js]: Had forgotten this in the original code :blush: ... who would have thought unit testing would find bugs!

File: test/async_test.js (hunk 2 above)
Comment (marcoscaceres) [inline test/async_test.js]: fixed

**Comment by marcoscaceres** [comment]
Squashed commit history.

**Comment by oyiptong** [comment]
R+

**Pull request merged**